db/db_impl/db_impl.cc

```diff
@@ -2341,14 +2346,19 @@ InternalIterator* DBImpl::NewInternalIterator(
     }
     internal_iter = merge_iter_builder.Finish(
         read_options.ignore_range_deletions ? nullptr : db_iter);
-    SuperVersionHandle* cleanup = new SuperVersionHandle(
-        this, &mutex_, super_version,
-        read_options.background_purge_on_iterator_cleanup ||
-            immutable_db_options_.avoid_unnecessary_blocking_io);
-    internal_iter->RegisterCleanup(CleanupSuperVersionHandle, cleanup, nullptr);
+    // Do not clean up the super version if super snapshot owns it
```

Comment:
Are we assuming that the super_version passed to NewInternalIterator is always the same as read_options.snapshot.sv_ here?

If so, how is that guarantee maintained?
If not, why can't we cleanup super_version here?

Reply:
Asking because `NewInternalIterator` is called in multiple places, but seems we only maintain the assumption in `DBImpl::NewIteratorImpl`.

Reply:
Yes, this is the assumption here, and you're right that it's not respected in some scenarios (like iterator refresh, although iterator refresh is also not supported if snapshot is non-nullptr). I wll add a check to only avoid cleaning up if `super_version` is not the same as `read_options.snapshot.sv_`.

Reply:
Addressed this comment in latest version of the code.